File: leetcode/java/longestConsecutive.java
/*
 * given an unsorted array of integers, find length of the longest consecutive elements sequence. implement it in time O(n)
 * */
import java.io.*;
import java.util.*;

public class longestConsecutive{
    /*
     * very smart and sophiscated structure design
     * */
    public int longestconsecutivesequence(int[] num){
        HashMap<Integer, Integer> mp = new HashMap<Integer, Integer>();    //Collection<Integer...>
        int maxlen = 1;
        int n = num.length;
        for(int i=0;i<n;i++){
            if(mp.containsKey(num[i]))    continue;
            mp.put(num[i], 1);
            int leftbound = 0;
            if(mp.containsKey(num[i]-1))    leftbound = mp.get(num[i]-1);
            int rightbound = 0;
            if(mp.containsKey(num[i]+1))    rightbound = mp.get(num[i]+1);
            int sum = 1 + leftbound + rightbound;
            mp.put(num[i]-leftbound, sum);
            mp.put(num[i]+rightbound, sum);
            if(sum > maxlen){
                maxlen = sum;
            }
        }
        return maxlen;
    }
    
    public int[] splitStr2IntArray(String str, String delimeters){
        StringTokenizer t = new StringTokenizer(str, delimeters);
        int[] arr = new int[t.countTokens()];
        for(int i=0;t.hasMoreTokens();i++){
            arr[i] = Integer.parseInt(t.nextToken().toString());
        }
        return arr;
    }
    public void test(){
        Scanner scan = new Scanner(System.in);
        while(true){
            System.out.println("input integer array:");
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            int[] num = splitStr2IntArray(str, " ,");
            System.out.println("length of longest consecutive sequence is " + longestconsecutivesequence(num));
        }
    }
    public static void main(String[] args){
        longestConsecutive lc = new longestConsecutive();
        lc.test();
    }
}

File: leetcode/src/longestConsecutive.cpp
/*
 * given an unsorted array of integers, find the length of the longest consecutive elements sequence.
 * complete it in time O(n)
 *
 * test case:
 * [100, 4, 200, 1, 3, 2], the longest consecutive sequence is [1,2,3,4], length is 4
 *
 * */

#include "../include/preliminary.h"
#include <unordered_map>

class Solution{

public:
    /*
     * hash_map<>(unordered_map<>) must be used
     * 1. how to enable map[i] present consecutive sequence length covering it ?
     * 2. how to process upper(right)/lower(left) bound ?
     * */
    int longestConsecutive_01(vector<int> &num){
        if(num.empty())    return 0;

        unordered_map<int, int> uii; // uii[k] tells longest consecutive width of which k is either leftbound or right bound

        int maxlen = 1;

        for(auto &val : num){
            // NOTE: operator[] is overused, while key not in, after operator[] is called, size increases by 1
            if(uii[val] != 0)    continue;

            uii[val] = 1;

            int leftbound = uii[val - 1], rightbound = uii[val + 1];

            // if leftbound = 0 and rightbound = 0, uii[val] = 1 remains
            uii[val - leftbound] = uii[val + rightbound] = 1 + leftbound + rightbound;

            maxlen = max(maxlen, 1 + leftbound + rightbound);
        }

        return maxlen;
    }

    /*
     * revise on solution1. it calls find() instead of operator[]
     * */
    int longestConsecutive_02(vector<int> &num){
        if(num.empty())    return 0;

        unordered_map<int, int> uii;

        int maxlen = 1;

        unordered_map<int, int>::const_iterator it = uii.end();

        for(auto &val : num){
            it = uii.find(val);
            if(it != uii.end())    continue;

            uii[val] = 1;

            int leftbound = 0;

            it = uii.find(val - 1);
            if(it != uii.end())    leftbound = it->second;

            int rightbound = 0;

            it = uii.find(val + 1);
            if(it != uii.end())    rightbound = it->second;

            uii[val - leftbound] = uii[val + rightbound] = 1 + leftbound + rightbound;

            maxlen = max(maxlen, 1 + leftbound + rightbound);
        }

        return maxlen;
    }
};

/* unit test is in ../cpp_unittest/longestConsecutive_unittest */
